Pull out TornadoFX configurations from AST parsing

ahinchman1 opened the pull request:
Big fat refactor!

- Pull out TornadoFX configurations out of KParser
- Add type aliases for better documentation of HashMap stores
- Add an interface and documentation for KParser
- Add basic unit tests for pure functions of KParser
- Various bits code cleanup

Commit:
wip - attempted pull out of TornadoFX configurations

Commit:
Looks like I've successfully pulled out TornadoFX configurations? Will need to allow custom configurations at the creation of KParser and make TornadoFX specific configurations not independent functions... maybe an object

Commit:
wip: test messes for both the UI and started KParsing testing :(

mattmoore commented on src/main/kotlin/com/github/ast/parser/TornadoFX.kt:
```diff
@@ -17,9 +24,114 @@ enum class NODES {
+ * TornadoFX specific:
+ *    Component breakdown meant to be saved and used for test generation
+ */
+fun saveComponentBreakdown(superClass: String, className: String, parser: KParser) {
```
If I'm understanding this correctly, this is kinda neat! Looks like you're able to pass this higher-order function in to specify that we only want to select TFX classes, excluding others.

NIT: Could be written as:

```suggestion
fun saveComponentBreakdown(superClass: String, className: String, parser: KParser) {
    val isTornadoFXView = listOf(COMPONENTS.Fragment.name, COMPONENTS.View.name)
            .contains(superClass)
    if (isTornadoFXView) parser.views[className] = TornadoFXView(className, superClass)
}
```

mattmoore commented on src/main/kotlin/com/github/ast/parser/TornadoFX.kt:
```diff
@@ -17,9 +24,114 @@ enum class NODES {
+ * TornadoFX specific:
+ *    Component breakdown meant to be saved and used for test generation
+ */
+fun saveComponentBreakdown(superClass: String, className: String, parser: KParser) {
```
If I'm understanding this correctly, this is kinda neat! Looks like you're able to pass this higher-order function in to specify that we only want to select TFX classes, excluding others.

NIT: Could be written as:

```suggestion
fun saveComponentBreakdown(superClass: String, className: String, parser: KParser) {
    val isTornadoFXView = listOf(COMPONENTS.Fragment.name, COMPONENTS.View.name)
            .contains(superClass)
    if (isTornadoFXView) parser.views[className] = TornadoFXView(className, superClass)
}
```

ahinchman1 replied:
Yeah, that's the idea! I really don't like how I did it without a total lack of structure though. I was just like, f*** it I'll rip it out and deal with it later.  Hmmm... I think this could work from eyeballing it.  It would be worth adding a test for this if we write it up this way.  I'll add your suggestion and maybe add testing for TornadoFX-specific configs as well!

mattmoore commented on src/main/kotlin/com/github/ast/parser/KParser.kt:
```diff
@@ -51,20 +54,16 @@ class KParser: Controller() {
         val classParents = ArrayList<String>()
-        val currentTFXView = TornadoFXView()
 
         gson.toJsonTree(file).asJsonObject
```
Looks like this isn't doing anything?

ahinchman1 replied:
Oh yeah I took this out. Thanks!

mattmoore commented on src/main/kotlin/com/github/ast/parser/NodeExtensions.kt:
```diff
@@ -0,0 +1,62 @@
+import com.google.gson.JsonArray
+import com.google.gson.JsonObject
+
+fun JsonObject.getType() = this.pieces().getObject(0).name()
```
Moving these into a separate file is great! Really helps clean up the KParser.

ahinchman1 replied:
Fr, as if that file isn't large enough already. Also, it works well as a set of independent functions!

Commit:
Major refactor. Created type alias for better documentation, added an interface for KParser for documentation and creating more flexible testing. Scattered cleanup and reorganization.

finished documentation and added a base set of unit tests for KParsing.

wip - attempted pull out of TornadoFX configurations

Looks like I've successfully pulled out TornadoFX configurations? Will need to allow custom configurations at the creation of KParser and make TornadoFX specific configurations not independent functions... maybe an object

wip: application is running but there is a bug where only one node is counted in the digraph

further refactoring.  MappedClassNodes appears to be erased for every iteration of the class set, so I will need to figure out how to preserve that hashmap

fixed the bug - I had been initializing KParser at every file *facepalm*

mattmoore commented on src/main/kotlin/com/github/hd/tornadofxsuite/view/Dialog.kt:
```diff
@@ -29,8 +29,8 @@ class Dialog : Fragment() {
                     listview<String> {
-                        if (scanner.mapClassViewNodes.size > 0) {
-                            scanner.mapClassViewNodes.forEach { (className, digraph) ->
+                        if (fxTestGenerator.scanner.mapClassViewNodes.size > 0) {
```
I don't think you need to check the size. I don't think `forEach` will run unless the size is greater than 0.

ahinchman1 replied:
lol Jonathan pointed out a similar thing for me previously. I have to double check with how that would work with HashMaps but I will likely apply your suggestion

Commit:
removed gradle files from before

removed .gradle/* files

put gradle back in

Commit:
minor code cleanup and test cleanup.  The imports doesn't seem to be compiling quite right for testing

Commit:
now have to revert kts because it's not working well with testing

Commit:
Fix gradle test stuff.

ahinchman1 committed:
Fix gradle test stuff?. (#30)

Fix gradle test stuff.

Commit:
Fixed file structure for testing and IntelliJ configuration

Co-authored-by: Matt Moore <mattmoore@carbonhelix.com>

Commit:
working version with spy, but I don't think I want that

Commit:
added another test for spying, but separated some independent functions out into a utils file to play with mocking later
